test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java

```diff
@@ -798,6 +800,17 @@ public static byte[] randomByteArrayOfLength(int size) {
         return bytes;
     }
 
+    public static BytesReference randomBytesReference(int length) {
+        final var slices = new ArrayList<BytesReference>();
+        var remaining = length;
+        while (remaining > 0) {
+            final var sliceLen = between(1, remaining);
+            slices.add(new BytesArray(randomByteArrayOfLength(sliceLen)));
+            remaining -= sliceLen;
+        }
+        return CompositeBytesReference.of(slices.toArray(BytesReference[]::new));
```

Comment:
Maybe we should generate the other `BytesReference` variants too?

Reply:
IMO this is enough - it means that we make `#iterator()` yield a fairly random selection of slice sizes, which is where I suspect the most bugs lurk. We'll also occasionally return a `BytesArray` in the case that there's only one slice.